ext/sodium/libsodium.c

```diff
@@ -0,0 +1,2682 @@
+							  &key, &key_len) == FAILURE) {
+		return;
+	}
+	if (ciphertext_len <= 0 || ciphertext_len >= SIZE_MAX) {
+		zend_throw_exception(sodium_exception_ce, "crypto_stream(): invalid length", 0);
+		return;
+	}
+	if (nonce_len != crypto_stream_NONCEBYTES) {
+		zend_throw_exception(sodium_exception_ce, "nonce should be CRYPTO_STREAM_NONCEBYTES bytes", 0);
+		return;
+	}
+	if (key_len != crypto_stream_KEYBYTES) {
+		zend_throw_exception(sodium_exception_ce, "key should be CRYPTO_STREAM_KEYBYTES bytes", 0);
+		return;
+	}
+	ciphertext = zend_string_alloc((size_t) ciphertext_len, 0);
+	if (crypto_stream((unsigned char *) ZSTR_VAL(ciphertext),
+					  (unsigned long long) ciphertext_len, nonce, key) != 0) {
+		zend_string_free(ciphertext);
+		zend_throw_exception(sodium_exception_ce, "crypto_stream(): internal error", 0);
+		return;
+	}
+	ZSTR_VAL(ciphertext)[ciphertext_len] = 0;
+
+	RETURN_STR(ciphertext);
+}
+
+PHP_FUNCTION(sodium_crypto_stream_xor)
+{
+	zend_string	  *ciphertext;
+	unsigned char *key;
+	unsigned char *msg;
+	unsigned char *nonce;
+	size_t		   ciphertext_len;
+	size_t		   key_len;
+	size_t		   msg_len;
+	size_t		   nonce_len;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "sss",
+							  &msg, &msg_len,
+							  &nonce, &nonce_len,
+							  &key, &key_len) == FAILURE) {
+		return;
+	}
+	if (nonce_len != crypto_stream_NONCEBYTES) {
+		zend_throw_exception(sodium_exception_ce, "nonce should be CRYPTO_STREAM_NONCEBYTES bytes", 0);
+		return;
+	}
+	if (key_len != crypto_stream_KEYBYTES) {
+		zend_throw_exception(sodium_exception_ce, "key should be CRYPTO_STREAM_KEYBYTES bytes", 0);
+		return;
+	}
+	ciphertext_len = msg_len;
+	ciphertext = zend_string_alloc((size_t) ciphertext_len, 0);
+	if (crypto_stream_xor((unsigned char *) ZSTR_VAL(ciphertext), msg,
+						  (unsigned long long) msg_len, nonce, key) != 0) {
+		zend_string_free(ciphertext);
+		zend_throw_exception(sodium_exception_ce, "crypto_stream_xor(): internal error", 0);
+		return;
+	}
+	ZSTR_VAL(ciphertext)[ciphertext_len] = 0;
+
+	RETURN_STR(ciphertext);
+}
+
+PHP_FUNCTION(sodium_crypto_pwhash_scryptsalsa208sha256)
+{
+	zend_string	  *hash;
+	unsigned char *salt;
+	char		  *passwd;
+	zend_long	   hash_len;
+	zend_long	   memlimit;
+	zend_long	   opslimit;
+	size_t		   passwd_len;
+	size_t		   salt_len;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "lssll",
+							  &hash_len,
+							  &passwd, &passwd_len,
+							  &salt, &salt_len,
+							  &opslimit, &memlimit) == FAILURE ||
+		hash_len <= 0 || hash_len >= SIZE_MAX ||
+		opslimit <= 0 || memlimit <= 0 || memlimit > SIZE_MAX) {
+		zend_throw_exception(sodium_exception_ce, "crypto_pwhash_scryptsalsa208sha256(): invalid parameters", 0);
+		return;
+	}
+	if (passwd_len <= 0) {
+		zend_error(E_WARNING, "empty password");
+	}
+	if (salt_len != crypto_pwhash_scryptsalsa208sha256_SALTBYTES) {
+		zend_throw_exception(sodium_exception_ce,
+				   "salt should be CRYPTO_PWHASH_SCRYPTSALSA208SHA256_SALTBYTES bytes",
```

Comment:
Suddenly the function name disappears from the messages. This should be consistent throughout the whole extension. As I wrote earlier, leaving it out is actually good enough, since the name of the function is part of the throwable's context anyways.